File: codegen/templates/python/dynamic/BaseAI.py
# -*- python -*-

from library import library

class BaseAI:
  """@brief A basic AI interface.

  This class implements most the code an AI would need to interface with the lower-level game code.
  AIs should extend this class to get a lot of builer-plate code out of the way
  The provided AI class does just that.
  """
  #\cond
  initialized = False
  iteration = 0
  runGenerator = None
  connection = None
  #\endcond
% for model in models:
%   if model.type == 'Model':
  ${lowercase(model.plural)} = []
%   endif
% endfor
  #\cond
  def startTurn(self):
% for model in models:
%   if model.type == 'Model':
    from GameObject import ${model.name}
%   endif
% endfor

% for model in models:
%   if model.type == 'Model':
    BaseAI.${lowercase(model.plural)} = [${model.name}(library.get${model.name}(self.connection, i)) for i in xrange(library.get${model.name}Count(self.connection))]
%   endif
% endfor

    if not self.initialized:
      self.initialized = True
      self.init()
    BaseAI.iteration += 1;
    if self.runGenerator:
      try:
        return self.runGenerator.next()
      except StopIteration:
        self.runGenerator = None
    r = self.run()
    if hasattr(r, '__iter__'):
      self.runGenerator = r
      return r.next()
    return r
  #\endcond
% for datum in globals:
  #\cond
  def get${capitalize(datum.name)}(self):
    return library.get${capitalize(datum.name)}(self.connection)
  #\endcond
  ##${datum.doc}
  ${datum.name} = property(get${capitalize(datum.name)})
% endfor
  def __init__(self, connection):
    self.connection = connection

File: codegen/templates/csharp/dynamic/BaseAI.cs
using System;
using System.Runtime.InteropServices;

/// \brief A basic AI interface.

///This class implements most the code an AI would need to interface with the lower-level game code.
///AIs should extend this class to get a lot of builer-plate code out of the way
///The provided AI class does just that.
public abstract class BaseAI
{
% for model in models:
%   if model.type == 'Model':
  public static ${model.name}[] ${lowercase(model.plural)};
%   endif
% endfor
  IntPtr connection;
  public static int iteration;
  bool initialized;

  public BaseAI(IntPtr c)
  {
    connection = c;
  }

  ///
  ///Make this your username, which should be provided.
  public abstract String username();
  ///
  ///Make this your password, which should be provided.
  public abstract String password();
  ///
  ///This is run on turn 1 before run
  public abstract void init();
  ///
  ///This is run every turn . Return true to end the turn, return false
  ///to request a status update from the server and then immediately rerun this function with the
  ///latest game status.
  public abstract bool run();

  ///
  ///This is run on after your last turn.
  public abstract void end();


  public bool startTurn()
  {
    iteration++;
    int count = 0;
% for model in models:
%   if model.type == 'Model':
    count = Client.get${model.name}Count(connection);
    ${lowercase(model.plural)} = new ${model.name}[count];
    for(int i = 0; i < count; i++)
    {
      ${lowercase(model.plural)}[i] = new ${model.name}(Client.get${model.name}(connection, i));
    }
%   endif
% endfor

    if(!initialized)
    {
      initialized = true;
      init();
    }
    return run();
  }


% for datum in globals:
  ///${datum.doc}
  public ${types[datum.type]} ${datum.name}()
  {
    ${types[datum.type]} value = Client.get${capitalize(datum.name)}(connection);
%   if datum.type is str:
    return Marshal.PtrToStringAuto(value);
%   else:
    return value;
%   endif
  }
% endfor
}
